clang/lib/CIR/Dialect/Transforms/CallConvLowering.cpp

```diff
@@ -23,56 +24,123 @@
 namespace mlir {
 namespace cir {
 
-//===----------------------------------------------------------------------===//
-// Rewrite Patterns
-//===----------------------------------------------------------------------===//
+FuncOp findFun(mlir::ModuleOp mod, llvm::StringRef name) {
+  FuncOp fun;
+  mod->walk([&](FuncOp f) {
+    if (f.getName() == name) {
+      fun = f;
+      return WalkResult::interrupt();
+    }
+    return WalkResult::advance();
+  });
 
-struct CallConvLoweringPattern : public OpRewritePattern<FuncOp> {
-  using OpRewritePattern<FuncOp>::OpRewritePattern;
+  return fun;
+}
 
-  LogicalResult matchAndRewrite(FuncOp op,
-                                PatternRewriter &rewriter) const final {
-    llvm::TimeTraceScope scope("Call Conv Lowering Pass", op.getSymName().str());
+bool isFuncPointerTy(mlir::Type typ) {
+  if (auto ptr = dyn_cast<PointerType>(typ))
+    return isa<FuncType>(ptr.getPointee());
+  return false;
+}
 
-    const auto module = op->getParentOfType<mlir::ModuleOp>();
+struct CallConvLowering {
+
+  CallConvLowering(LowerModule &mod, mlir::PatternRewriter &rw,
+                   mlir::TypeConverter &converter)
+      : lowerModule(mod), rewriter(rw), typeConverter(converter) {}
+
+  void lower(Operation *op) {
+    rewriter.setInsertionPoint(op);
+    if (auto fun = dyn_cast<FuncOp>(op))
+      lower(fun);
```

Comment:
I'd prefer if we don't overload in the "visitor pattern style", perhaps create one single method for each: `lowerFuncOp`, `lowerAllocaOp`, etc.